notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,458 @@
+    "    positive_indices = {}\n",
+    "    negative_indices = {}\n",
+    "    recidivism_count = {}\n",
+    "    recidivism_pct = {}\n",
+    "    groups = np.unique(sensitive_features.values)\n",
+    "    n_groups = len(groups)\n",
+    "    max_group_length = max([len(group) for group in groups])\n",
+    "    color = cm.rainbow(np.linspace(0,1,n_groups))\n",
+    "    x_tick_labels_basic = []\n",
+    "    x_tick_labels_by_label = []\n",
+    "    for index, group in enumerate(groups):\n",
+    "        indices[group] = sensitive_features.index[sensitive_features == group]\n",
+    "        recidivism_count[group] = sum(y_pred[indices[group]])\n",
+    "        recidivism_pct[group] = recidivism_count[group]/len(indices[group])\n",
+    "        print(\"P[recidivism predicted | {}]                {}= {}\".format(group, \" \"*(max_group_length-len(group)), recidivism_pct[group]))\n",
+    "    \n",
+    "        plt.bar(index + 1, recidivism_pct[group], color=color[index])\n",
+    "        x_tick_labels_basic.append(group)\n",
+    "    \n",
+    "        if y is not None:\n",
+    "            positive_indices[group] = sensitive_features.index[(sensitive_features == group) & (y == 1)]\n",
+    "            negative_indices[group] = sensitive_features.index[(sensitive_features == group) & (y == 0)]\n",
+    "            prob_1 = sum(y_pred[positive_indices[group]])/len(positive_indices[group])\n",
+    "            prob_0 = sum(y_pred[negative_indices[group]])/len(negative_indices[group])\n",
+    "            print(\"P[recidivism predicted | {}, recidivism]    {}= {}\".format(group, \" \"*(max_group_length-len(group)) , prob_1))\n",
+    "            print(\"P[recidivism predicted | {}, no recidivism] {}= {}\".format(group, \" \"*(max_group_length-len(group)), prob_0))\n",
+    "\n",
+    "            plt.bar(n_groups + 1 + 2 * index, prob_1, color=color[index])\n",
+    "            plt.bar(n_groups + 2 + 2 * index, prob_0, color=color[index])\n",
+    "            x_tick_labels_by_label.extend([\"{} recidivism\".format(group), \"{} no recidivism\".format(group)])\n",
+    "    \n",
+    "    x_tick_labels = x_tick_labels_basic + x_tick_labels_by_label\n",
+    "    plt.xticks(range(1, len(x_tick_labels)+1), x_tick_labels, rotation=45, horizontalalignment=\"right\")\n",
+    "        \n",
+    "show_proportions(X_train, sensitive_features_train, y_train, description=\"original training data:\", plot_row_index=1)\n",
+    "show_proportions(X_train, sensitive_features_train, unconstrained_predictor.predict(X_train), y_train, description=\"fairness-unaware prediction on training data:\", plot_row_index=2)\n",
+    "show_proportions(X_test, sensitive_features_test, y_test, description=\"original test data:\", plot_row_index=3)\n",
+    "show_proportions(X_test, sensitive_features_test, unconstrained_predictor.predict(X_test), y_test, description=\"fairness-unaware prediction on test data:\", plot_row_index=4)\n",
+    "plt.show()"
```

Comment:
Can these plots be done in the new widget?

Reply:
Also, yes there's a difference, but what are we supposed to take from this? I'm also not sure that the columns in the graphs are ordered usefully for understanding them. But again, I'm not sure exactly what I'm supposed to be taking away from this

Reply:
The widget is too large to show this in a coherent way. Also, the widget doesn't show me all this data. This is relevant because it's what we optimize for with EO.

Reply:
As we discovered they cannot be done in the widget, so I'll leave them here for now. The goal is to eventually get rid of them if the widget gets that functionality